File: pj_backend/app/Models/Shipment.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Concerns\HasUuids;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\Relations\HasOne;

class Shipment extends Model
{
    use HasUuids;

    protected $fillable = [
        'customer_id',
        'driver_id',
        'last_rejected_driver_id',
        'last_assignment_rejection_reason',
        'last_assignment_rejected_at',
        'origin',
        'destination',
        'product_platform',
        'product_url',
        'product_external_id',
        'product_title',
        'product_image_url',
        'product_price',
        'product_color',
        'product_size',
        'product_metadata',
        'transit_countries',
        'weight_kg',
        'size',
        'category_id',
        'vehicle_type_id',
        'price_breakdown',
        'total_price',
        'estimated_delivery_days',
        'status',
        'delivery_confirmed_at',
    ];

    protected $casts = [
        'transit_countries' => 'array',
        'product_metadata' => 'array',
        'price_breakdown' => 'array',
        'product_price' => 'float',
        'weight_kg' => 'float',
        'total_price' => 'float',
        'estimated_delivery_days' => 'integer',
        'customer_id' => 'integer',
        'category_id' => 'integer',
        'vehicle_type_id' => 'integer',
        'driver_id' => 'integer',
        'last_rejected_driver_id' => 'integer',
        'last_assignment_rejected_at' => 'datetime',
        'delivery_confirmed_at' => 'datetime',
    ];

    public function customer(): BelongsTo
    {
        return $this->belongsTo(User::class, 'customer_id');
    }

    public function driver(): BelongsTo
    {
        return $this->belongsTo(User::class, 'driver_id');
    }

    public function lastRejectedDriver(): BelongsTo
    {
        return $this->belongsTo(User::class, 'last_rejected_driver_id');
    }

    public function category(): BelongsTo
    {
        return $this->belongsTo(Category::class);
    }

    public function vehicleType(): BelongsTo
    {
        return $this->belongsTo(VehicleType::class);
    }

    public function report(): HasOne
    {
        return $this->hasOne(Report::class);
    }

    public function notifications(): HasMany
    {
        return $this->hasMany(Notification::class);
    }
}

File: packages/pj_domain/lib/src/models/shipment.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'category.dart';
import 'report.dart';
import 'user.dart';
import 'vehicle_type.dart';

part 'shipment.freezed.dart';
part 'shipment.g.dart';

enum ShipmentStatus {
  pending,
  @JsonValue('in_transit')
  inTransit,
  delivered,
  reported,
}

@freezed
class Shipment with _$Shipment {
  const factory Shipment({
    required String id,
    @JsonKey(name: 'customer_id') required int customerId,
    @JsonKey(name: 'driver_id') int? driverId,
    required String origin,
    required String destination,
    @JsonKey(name: 'product_platform') String? productPlatform,
    @JsonKey(name: 'product_url') String? productUrl,
    @JsonKey(name: 'product_external_id') String? productExternalId,
    @JsonKey(name: 'product_title') String? productTitle,
    @JsonKey(name: 'product_image_url') String? productImageUrl,
    @JsonKey(name: 'product_price') double? productPrice,
    @JsonKey(name: 'product_color') String? productColor,
    @JsonKey(name: 'product_size') String? productSize,
    @JsonKey(name: 'product_metadata') Map<String, dynamic>? productMetadata,
    @JsonKey(name: 'transit_countries') List<String>? transitCountries,
    @JsonKey(name: 'weight_kg') double? weightKg,
    @JsonKey(name: 'size') String? size,
    @JsonKey(name: 'category_id') required int categoryId,
    @JsonKey(name: 'vehicle_type_id') required int vehicleTypeId,
    @JsonKey(name: 'total_price') required double totalPrice,
    @JsonKey(name: 'estimated_delivery_days')
    required int estimatedDeliveryDays,
    @Default(ShipmentStatus.pending) ShipmentStatus status,
    @JsonKey(name: 'delivery_confirmed_at') DateTime? deliveryConfirmedAt,
    Category? category,
    @JsonKey(name: 'vehicle_type') VehicleType? vehicleType,
    User? customer,
    User? driver,
    Report? report,
    @JsonKey(name: 'price_breakdown') Map<String, dynamic>? priceBreakdown,
  }) = _Shipment;

  factory Shipment.fromJson(Map<String, dynamic> json) =>
      _$ShipmentFromJson(json);
}
